File: scripts/clusters.jl
if size(ARGS, 1) < 3
    error("Please provide generating.json, exploring.json and snap shots.")
end
import LightGraphs; LG = LightGraphs
import TikzPictures; TP = TikzPictures
import TikzGraphs; TG = TikzGraphs
import PLMC
import JSON

generatingData = JSON.parsefile(ARGS[1]; dicttype=Dict, use_mmap=true)
i_box = 1 # data will be overridden when reading snap shots
periodicBox = PLMC.newBox(i_box, generatingData)
components = PLMC.newComponents(i_box, generatingData)
exploringData = JSON.parsefile(ARGS[2]; dicttype=Dict, use_mmap=true)
maximumEdgeDistance = exploringData["Clusters"]["maximum distance"]

for i_snap=1:size(ARGS[3:end], 1)

    open(ARGS[2+i_snap]) do coordinates
        PLMC.readDipolesCoordinates(periodicBox, components, coordinates)
    end

    for component in components
        graph = PLMC.newDiGraph(periodicBox, component, maximumEdgeDistance)

        #=
        println("cycles = ", collect(nx.simple_cycles(py_graph)))
        #println("wcc = ", collect(nx.weakly_connected_components(py_graph)))
        clusters = nx.weakly_connected_component_subgraphs(py_graph)
        for cluster in clusters
            if !isempty(collect(nx.simple_cycles(cluster)))
                println("cluster.cycles = ", collect(nx.simple_cycles(cluster)))
            end
        end
        =#

        #TP.save(TP.PDF("graph_$(i_snap)_$(i_component).PDF"), TG.plot(graph))
        clusters = LG.weakly_connected_components(graph)
        for cluster in clusters
            if size(cluster, 1) > 1
                println(cluster)
            end
        end
        #=
        for i_cluster = 1:size(clusters, 1)
            if size(clusters[i_cluster], 1) > 2
                sub_graph = LG.induced_subgraph(graph, clusters[i_cluster])
                println(i_cluster,
                    ": nv = ", LG.nv(sub_graph),
                    " is_cyclic: ", LG.is_cyclic(sub_graph))
                TP.save(TP.PDF("graph_$(i_snap)_$(i_component)_$(i_cluster).PDF"),
                    TG.plot(sub_graph))
                #TP.save(TP.PDF("graph_$(i_snap)_$(i_component)_$(i_cluster)_condensation.pdf"),
                #    TG.plot(LG.condensation(sub_graph), TG.Layouts.Spring()))
            end
        end
        =#
    end

end

File: scripts/clusters.py
import argparse
import json
import numpy as np
import plmc
import networkx as nx
import matplotlib.pyplot as plt

parser = argparse.ArgumentParser()
parser.add_argument("generatingFile")
parser.add_argument("exploringFile")
parser.add_argument("snapshots", type=argparse.FileType('r'), nargs='*')

with open(parser.parse_args().generatingFile) as file:
    generatingData = json.load(file)
    i_box = 0
    periodicBox = plmc.newBox(i_box, generatingData)
    components = plmc.newComponents(i_box, generatingData)

with open(parser.parse_args().exploringFile) as file:
    exploringData = json.load(file)
    maximumEdgeDistance = exploringData["Dipolar Graph"]["maximum distance"]

graph = nx.DiGraph()
for snapshot in parser.parse_args().snapshots:
    
    periodicBox.size = np.array([float(boxSize_i) for boxSize_i in snapshot.readline().split()[2:]])
    numsParticles = [int(numParticle) for numParticle in snapshot.readline().split()[2:]]
    for i_component in range(len(components)):
        components[i_component].num = numsParticles[i_component]
    
    snapshot.readline() #header
    for component in components:
        if component.isDipolar:
            component.positions = np.zeros([component.num, 3])
            component.orientations = np.zeros([component.num, 3])
            for i_particle in range(component.num):
                coordinates_i = [float(coordinates_ij) for coordinates_ij in \
                    snapshot.readline().split()[1:]]
                component.positions[i_particle] = coordinates_i[0:3]
                component.orientations[i_particle] = coordinates_i[3:6] 
        else:
            for i_particle in range(component.num):
                snapshot.readline() # apolar coordinates

    for component in components:
        if component.isDipolar:
            print(component)
            for i_particle in range(component.num):
                for j_particle in range(i_particle):
                    vector_ij = periodicBox.vector(component.positions[i_particle], \
                        component.positions[j_particle])
                    if np.linalg.norm(vector_ij) < maximumEdgeDistance and \
                        plmc.dipolarEnergyIsNegative(vector_ij, component.orientations[i_particle],\
                            component.orientations[j_particle]):
                        if np.dot(vector_ij, component.orientations[i_particle]) > 0:
                            graph.add_edge(i_particle, j_particle)
                        else:
                            graph.add_edge(j_particle, i_particle)
            
            weakly_connected_components = nx.weakly_connected_component_subgraphs(graph)
            for sub_graphs in weakly_connected_components:
                #print("sub_graphs", list(sub_graphs))
                if len(list(nx.simple_cycles(sub_graphs))) > 0:
                    print("simple_cycles", list(nx.simple_cycles(sub_graphs)))
                    nx.draw_networkx(sub_graphs)
                    plt.show()